File: test/pattern/test_prototype.cpp
#include <gtest/gtest.h>
#include "pattern/prototype.hpp"

using namespace pattern;

// ---------------------------------------------------------------------------
// Basic clone tests
// ---------------------------------------------------------------------------
TEST(Prototype, CloneProducesNewObject)
{
    Circle original("red", 5.0);
    auto cloned = original.clone();
    EXPECT_NE(&original, cloned.get());
}

TEST(Prototype, ClonePreservesType)
{
    Circle original("blue", 3.0);
    auto cloned = original.clone();
    EXPECT_EQ("Circle", cloned->type());
}

TEST(Prototype, ClonePreservesColor)
{
    Circle original("green", 4.0);
    auto cloned = original.clone();
    EXPECT_EQ("green", cloned->color());
}

TEST(Prototype, CloneCirclePreservesRadius)
{
    Circle original("red", 7.5);
    auto cloned = original.clone();
    Circle* c = dynamic_cast<Circle*>(cloned.get());
    ASSERT_NE(nullptr, c);
    EXPECT_DOUBLE_EQ(7.5, c->radius());
}

TEST(Prototype, CloneRectanglePreservesDimensions)
{
    Rectangle original("yellow", 10.0, 20.0);
    auto cloned = original.clone();
    Rectangle* r = dynamic_cast<Rectangle*>(cloned.get());
    ASSERT_NE(nullptr, r);
    EXPECT_DOUBLE_EQ(10.0, r->width());
    EXPECT_DOUBLE_EQ(20.0, r->height());
}

// ---------------------------------------------------------------------------
// Independence: mutating clone does not affect original
// ---------------------------------------------------------------------------
TEST(Prototype, MutatingCloneDoesNotAffectOriginal)
{
    Circle original("red", 5.0);
    auto cloned = original.clone();
    cloned->setColor("blue");
    EXPECT_EQ("red",  original.color());
    EXPECT_EQ("blue", cloned->color());
}

TEST(Prototype, MutatingCircleRadiusDoesNotAffectOriginal)
{
    Circle original("red", 5.0);
    auto cloned = original.clone();
    Circle* c = dynamic_cast<Circle*>(cloned.get());
    c->setRadius(99.0);
    EXPECT_DOUBLE_EQ(5.0,  original.radius());
    EXPECT_DOUBLE_EQ(99.0, c->radius());
}

// ---------------------------------------------------------------------------
// Registry tests
// ---------------------------------------------------------------------------
TEST(Prototype, RegistryReturnsClonesNotOriginals)
{
    ShapeRegistry reg;
    reg.add("circle", std::unique_ptr<Shape>(new Circle("red", 1.0)));

    auto a = reg.get("circle");
    auto b = reg.get("circle");
    EXPECT_NE(a.get(), b.get());
}

TEST(Prototype, RegistryPreservesProperties)
{
    ShapeRegistry reg;
    reg.add("rect", std::unique_ptr<Shape>(new Rectangle("blue", 4.0, 8.0)));

    auto s = reg.get("rect");
    EXPECT_EQ("blue",      s->color());
    EXPECT_EQ("Rectangle", s->type());
}

TEST(Prototype, RegistryThrowsOnMissingKey)
{
    ShapeRegistry reg;
    EXPECT_THROW(reg.get("nonexistent"), std::out_of_range);
}

TEST(Prototype, RegistryHas)
{
    ShapeRegistry reg;
    reg.add("circle", std::unique_ptr<Shape>(new Circle("red", 1.0)));
    EXPECT_TRUE(reg.has("circle"));
    EXPECT_FALSE(reg.has("triangle"));
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: scripts/test_pattern/test_prototype.py
"""Tests for the Prototype pattern."""

import sys
import os
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'pattern'))
from prototype import Circle, Rectangle, ShapeRegistry


class TestPrototype(unittest.TestCase):

    # ------------------------------------------------------------------
    # Basic clone
    # ------------------------------------------------------------------
    def test_clone_produces_new_object(self):
        c = Circle("red", 5.0)
        self.assertIsNot(c, c.clone())

    def test_clone_preserves_type(self):
        c = Circle("blue", 3.0)
        self.assertEqual("Circle", c.clone().type())

    def test_clone_preserves_color(self):
        c = Circle("green", 4.0)
        self.assertEqual("green", c.clone().color)

    def test_clone_circle_preserves_radius(self):
        c = Circle("red", 7.5)
        clone = c.clone()
        self.assertAlmostEqual(7.5, clone.radius)

    def test_clone_rectangle_preserves_dimensions(self):
        r = Rectangle("yellow", 10.0, 20.0)
        clone = r.clone()
        self.assertAlmostEqual(10.0, clone.width)
        self.assertAlmostEqual(20.0, clone.height)

    # ------------------------------------------------------------------
    # Independence
    # ------------------------------------------------------------------
    def test_mutating_clone_does_not_affect_original(self):
        c = Circle("red", 5.0)
        clone = c.clone()
        clone.color = "blue"
        self.assertEqual("red",  c.color)
        self.assertEqual("blue", clone.color)

    def test_mutating_radius_does_not_affect_original(self):
        c = Circle("red", 5.0)
        clone = c.clone()
        clone.radius = 99.0
        self.assertAlmostEqual(5.0, c.radius)
        self.assertAlmostEqual(99.0, clone.radius)

    # ------------------------------------------------------------------
    # Registry
    # ------------------------------------------------------------------
    def test_registry_returns_clones_not_originals(self):
        reg = ShapeRegistry()
        reg.add("circle", Circle("red", 1.0))
        a = reg.get("circle")
        b = reg.get("circle")
        self.assertIsNot(a, b)

    def test_registry_preserves_properties(self):
        reg = ShapeRegistry()
        reg.add("rect", Rectangle("blue", 4.0, 8.0))
        s = reg.get("rect")
        self.assertEqual("blue",      s.color)
        self.assertEqual("Rectangle", s.type())

    def test_registry_raises_on_missing_key(self):
        reg = ShapeRegistry()
        with self.assertRaises(KeyError):
            reg.get("nonexistent")

    def test_registry_contains(self):
        reg = ShapeRegistry()
        reg.add("circle", Circle("red", 1.0))
        self.assertIn("circle", reg)
        self.assertNotIn("triangle", reg)


if __name__ == "__main__":
    unittest.main()
